### PR TITLE
Feat/search after

### PR DESCRIPTION
**Pull request by maxserras**
Added the "from_id" parameter for the rb.load() method, the search and stream  endpoints in FastAPI. 

Now all the results from rb.load and the search over the datasets are sorted by record identifier. 

This feature allows to iterate over large datasets (+1M samples) without the need of loading everything in one go or do hacky stuff. Also, this makes easier to integrate with dataloaders to create batches for training models without the need of loading the whole dataset.

### CODE REVIEW

File: src/rubrix/client/sdk/text2text/api.py [hunk 1]
```diff
@@ -21,22 +21,25 @@
 from rubrix.client.sdk.commons.models import ErrorMessage, HTTPValidationError, Response
 from rubrix.client.sdk.text2text.models import Text2TextQuery, Text2TextRecord
 
+from src.rubrix.client.sdk.commons.api import build_param_dict
```
Comment (frascuchon) [inline src/rubrix/client/sdk/text2text/api.py]:
```suggestion
from rubrix.client.sdk.commons.api import build_param_dict
```

File: src/rubrix/client/sdk/text_classification/api.py [hunk 2]
```diff
@@ -27,24 +27,27 @@
     TextClassificationRecord,
 )
 
+from src.rubrix.client.sdk.commons.api import build_param_dict
```
Comment (frascuchon) [inline src/rubrix/client/sdk/text_classification/api.py]:
```suggestion
from rubrix.client.sdk.commons.api import build_param_dict
```

File: src/rubrix/client/sdk/token_classification/api.py [hunk 3]
```diff
@@ -25,26 +25,29 @@
     TokenClassificationRecord,
 )
 
+from src.rubrix.client.sdk.commons.api import build_param_dict
```
Comment (frascuchon) [inline src/rubrix/client/sdk/token_classification/api.py]:
```suggestion
from rubrix.client.sdk.commons.api import build_param_dict
```

File: src/rubrix/server/daos/records.py [hunk 4]
```diff
@@ -287,10 +303,19 @@ def scan_dataset(
...
+            # Scroll method does not accept read_after, thus, this case is handled as a search
+            es_query["search_after"] = [id_from]
+            results = self._es.search(index=dataset_records_index(dataset.id), query=es_query, size=limit)
+            hits = results["hits"]
+            docs = hits["hits"]
+
+        else:
+            docs = self._es.list_documents(
+                dataset_records_index(dataset.id), query=es_query,
+            )
```
Comment (frascuchon) [inline src/rubrix/server/daos/records.py:308]: Could we use the `search` method always for dataset loading? That means that the `rb.load` will fetch records always in batches, and can improve the client's overall performance. 

We avoid using index scrolling, what is a recommendation from the [official docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/scroll-api.html)

File: src/rubrix/server/apis/v0/handlers/text2text.py [hunk 5]
```diff
@@ -130,6 +130,7 @@ def bulk_records(
 )
 def search_records(
     name: str,
+    id_from: Optional[str] = None,
```
Comment (frascuchon) [inline src/rubrix/server/apis/v0/handlers/text2text.py]: Since it doesn't affect the loading of datasets with `rb.load`, I prefer to leave this parameter out of the search for now. 

The motivation is that we should solve the problem when searching with a sort provided to the API. With the proposed implementation, provided sorting has no effect and that worries me. 

It is an issue that can be resolved, but it is not easy. That's why I prefer to approach it independently.

File: src/rubrix/server/apis/v0/handlers/text_classification.py [hunk 6]
```diff
@@ -146,6 +146,7 @@ async def bulk_records(
 )
 def search_records(
     name: str,
+    id_from: Optional[str] = None,
```
Comment (frascuchon) [inline src/rubrix/server/apis/v0/handlers/text_classification.py]: Ditto

File: src/rubrix/server/apis/v0/handlers/token_classification.py [hunk 7]
```diff
@@ -141,6 +141,7 @@ async def bulk_records(
 )
 def search_records(
     name: str,
+    id_from: Optional[str] = None,
```
Comment (frascuchon) [inline src/rubrix/server/apis/v0/handlers/token_classification.py]: Ditto

File: src/rubrix/server/apis/v0/models/text2text.py [hunk 8]
```diff
@@ -215,6 +215,7 @@ class Text2TextSearchRequest(BaseModel):
 
     query: Text2TextQuery = Field(default_factory=Text2TextQuery)
     sort: List[SortableField] = Field(default_factory=list)
+    id_from: Optional[str] = None
```
Comment (frascuchon) [inline src/rubrix/server/apis/v0/models/text2text.py]: I think we can drop this field, since we're passing the `id_from` as id param instead of a body attibute

File: src/rubrix/server/apis/v0/models/text_classification.py [hunk 9]
```diff
@@ -531,6 +531,7 @@ class TextClassificationSearchRequest(BaseModel):
 
     query: TextClassificationQuery = Field(default_factory=TextClassificationQuery)
     sort: List[SortableField] = Field(default_factory=list)
+    id_from: Optional[str] = None
```
Comment (frascuchon) [inline src/rubrix/server/apis/v0/models/text_classification.py]: Ditto

File: src/rubrix/server/apis/v0/models/token_classification.py [hunk 10]
```diff
@@ -425,6 +425,7 @@ class TokenClassificationSearchRequest(BaseModel):
 
     query: TokenClassificationQuery = Field(default_factory=TokenClassificationQuery)
     sort: List[SortableField] = Field(default_factory=list)
+    id_from: Optional[str] = None
```
Comment (frascuchon) [inline src/rubrix/server/apis/v0/models/token_classification.py]: Ditto

File: src/rubrix/server/services/search/service.py [hunk 11]
```diff
@@ -60,6 +60,7 @@ def search(
         size: int = 100,
         exclude_metrics: bool = True,
         metrics: Optional[List[BaseMetric]] = None,
+        id_from: Optional[str] = None
```
Comment (frascuchon) [inline src/rubrix/server/services/search/service.py]: As I said before, we can wait to apply changes to the search

File: src/rubrix/server/services/text2text.py [hunk 12]
```diff
@@ -86,6 +86,7 @@ def search(
         sort_by: List[SortableField],
         record_from: int = 0,
         size: int = 100,
+        id_from: Optional[str] = None,
```
Comment (frascuchon) [inline src/rubrix/server/services/text2text.py]: Ditto

File: src/rubrix/server/services/text_classification.py [hunk 13]
```diff
@@ -97,6 +97,7 @@ def search(
         sort_by: List[SortableField],
         record_from: int = 0,
         size: int = 100,
+        id_from: Optional[str] = None,
```
Comment (frascuchon) [inline src/rubrix/server/services/text_classification.py]: Ditto

**Review by frascuchon — COMMENTED** [review]
It looks great to me. Just a few comments and some reflections about changes on the search endpoint.

File: src/rubrix/server/elasticseach/client_wrapper.py [hunk 14]
```diff
@@ -102,7 +102,8 @@ def list_documents(
         A sequence of documents resulting from applying the query on the index
 
         """
-        return es_scan(self.__client__, query=query or {}, index=index)
+        # Preserve order is used to ensure that the samples are sorted by identifier
+        return es_scan(self.__client__, query=query or {}, index=index, preserve_order=True)
```
Comment (frascuchon) [inline src/rubrix/server/elasticseach/client_wrapper.py]: From elasticsearch client documentation:

> preserve_order – don’t set the search_type to scan - this will cause the scroll to paginate with preserving the order. Note that this can be an extremely expensive operation and can easily lead to unpredictable results, use with caution.

I'm a bit unquiet about the performance impact caused by activating the `preserve_order` flag. If I understand well, this must be set up to fetch the proper last id for the next batch in `rb.load`, right? But, when no limit is applied, we will return the whole dataset, so no need for order preservation, right?

So, for those cases where a limit is provided, I suggest using the search method instead, sorted by the `id` field. Anyway, the better solution is to implement an alternative helper to `es_scan` using automatically the `search_after` attribute. Let me think about a solution that could fit in this PR

Comment (frascuchon) [inline src/rubrix/server/elasticseach/client_wrapper.py]: I've been doing some tests and I think this should generally work. We avoid using scroll and use the `search_after` mechanism to iterate the whole dataset:

```suggestion
        size = size or 1000
        query = query.copy() or {}
        query["sort"] = [{"_id": {"order": "asc"}}] # Force sorting by id
        response = self.__client__.search(index=index, body=query, size=size)
        while response["hits"]["hits"]:
            for hit in response["hits"]["hits"]:
                yield hit

            last_id = hit["_id"]
            query["search_after"] = [last_id]
            response = self.__client__.search(index=index, body=query, size=size)
```

File: src/rubrix/server/daos/records.py (hunk 4 above)
Comment (maxserras) [reply src/rubrix/server/daos/records.py:308]: Agree :D

File: src/rubrix/server/daos/records.py [hunk 15]
```diff
@@ -181,6 +181,7 @@ def search_records(
         record_from: int = 0,
         exclude_fields: List[str] = None,
         highligth_results: bool = True,
+        id_from: Optional[str] = None,
```
Comment (frascuchon) [inline src/rubrix/server/daos/records.py]: We should revert this change, and keep search_records as before

File: src/rubrix/server/daos/records.py [hunk 16]
```diff
@@ -208,16 +213,21 @@ def search_records(
         aggregation_requests = (
             {**(search.aggregations or {})} if compute_aggregations else {}
         )
-
-        sort_config = self.__normalize_sort_config__(records_index, sort=search.sort)
-
```
Comment (frascuchon) [inline src/rubrix/server/daos/records.py]: ditto

File: src/rubrix/server/daos/records.py [hunk 17]
```diff
@@ -208,16 +213,21 @@ def search_records(
...
         es_query = {
             "_source": {"excludes": exclude_fields or []},
             "from": record_from,
             "query": search.query or {"match_all": {}},
-            "sort": sort_config,
```
Comment (frascuchon) [inline src/rubrix/server/daos/records.py]: ditto

File: src/rubrix/server/daos/records.py [hunk 18]
```diff
@@ -208,16 +213,21 @@ def search_records(
...
         }
+
+        if not id_from:
+            sort_config = self.__normalize_sort_config__(records_index, sort=search.sort)
+        else:
+            sort_config = [{"id": {"order": "asc"}}]
+            search_after = [id_from]
+            es_query["search_after"] = search_after
+        es_query["sort"] = sort_config
+
```
Comment (frascuchon) [inline src/rubrix/server/daos/records.py]: ditto

File: src/rubrix/server/elasticseach/client_wrapper.py [hunk 19]
```diff
@@ -102,7 +102,17 @@ def list_documents(
         A sequence of documents resulting from applying the query on the index
 
         """
-        return es_scan(self.__client__, query=query or {}, index=index)
+        size = size or 1000
```
Comment (frascuchon) [inline src/rubrix/server/elasticseach/client_wrapper.py:107]: `size` must be defined as an argument. Otherwise, it fails on run:


```python
def list_documents(
        self, index: str, query: Dict[str, Any] = None, size: Optional[int] = None
    ) -> Iterable[Dict[str, Any]]:
```

**Review by frascuchon — APPROVED** [review]

**Comment by bilykigor** [comment]
Screen Shot 2022-09-08 at 1 55 35 PM

Seems id_from not working in 0.17.0

**Comment by frascuchon** [comment]
Hi @bilykigor, and thanks for reaching out.

Are you using v0.17.0 also for the server?
